cranelift-interpreter/src/interpreter.rs

```diff
@@ -0,0 +1,295 @@
+use cranelift_codegen::ir::condcodes::IntCC;
+use cranelift_codegen::ir::immediates::Imm64;
+use cranelift_codegen::ir::{
+    Ebb, FuncRef, Function, Inst, InstructionData, InstructionData::*, Opcode, Opcode::*,
+    Value as ValueRef, ValueList,
+};
+use log::debug;
+use std::ops::{Add, Sub};
+use thiserror::Error;
+
+/// The valid control flow states.
+pub enum ControlFlow {
+    Continue,
+    ContinueAt(Ebb, Vec<ValueRef>),
+    Return(Vec<Value>),
+}
+
+/// The ways interpretation can fail.
+#[derive(Error, Debug)]
+pub enum Trap {
+    #[error("unknown trap")]
+    Unknown,
+    #[error("invalid type for {1}: expected {0}")]
+    InvalidType(String, ValueRef),
+    #[error("reached an unreachable statement")]
+    Unreachable,
+    #[error("invalid control flow: {0}")]
+    InvalidControlFlow(String),
+    #[error("invalid function reference: {0}")]
+    InvalidFunctionReference(FuncRef),
+    #[error("invalid function name: {0}")]
+    InvalidFunctionName(String),
+}
+
+/// The Cranelift interpreter; it contains immutable elements such as the function environment and
+/// implements the Cranelift IR semantics.
+#[derive(Default)]
+pub struct Interpreter {
+    pub env: Environment,
+}
+
+impl Interpreter {
+    pub fn new(env: Environment) -> Self {
+        Self { env }
+    }
+
+    pub fn call_by_name(&self, func_name: &str, arguments: &[Value]) -> Result<ControlFlow, Trap> {
+        let func_ref = self
+            .env
+            .index_of(func_name)
+            .ok_or_else(|| Trap::InvalidFunctionName(func_name.to_string()))?;
+        self.call_by_index(func_ref, arguments)
+    }
+
+    pub fn call_by_index(
+        &self,
+        func_ref: FuncRef,
+        arguments: &[Value],
+    ) -> Result<ControlFlow, Trap> {
+        match self.env.get_by_func_ref(func_ref) {
+            None => Err(Trap::InvalidFunctionReference(func_ref)),
+            Some(func) => self.call(func, arguments),
+        }
+    }
+
+    fn call(&self, function: &Function, arguments: &[Value]) -> Result<ControlFlow, Trap> {
+        debug!("Call: {}({:?})", function.name, arguments);
+        let first_ebb = function.layout.ebbs().next().expect("to have a first ebb");
+        let parameters = function.dfg.ebb_params(first_ebb);
+        let mut frame = Frame::new(function).with_parameters(parameters, arguments);
+        let result = self.block(&mut frame, first_ebb);
+        result
+    }
+
+    fn block(&self, frame: &mut Frame, ebb: Ebb) -> Result<ControlFlow, Trap> {
+        debug!("Block: {}", ebb);
+        for inst in frame.function.layout.ebb_insts(ebb) {
+            match self.inst(frame, inst)? {
+                ControlFlow::Continue => continue,
+                ControlFlow::ContinueAt(ebb, old_names) => {
+                    let new_names = frame.function.dfg.ebb_params(ebb);
+                    frame.rename(&old_names, new_names);
+                    return self.block(frame, ebb); // TODO check that TCO happens
```

Comment:
The rebased commits have `block` instead of `ebb`. My TCO comment (now removed) was in hopes that Rust was somehow smart enough to avoid recursive calls... is there an easy way to check if that is happening?